src/charm_state.py

```diff
@@ -846,23 +846,37 @@ def _check_immutable_config_change(
         prev_state = json.loads(json_data)
         logger.info("Previous charm state: %s", prev_state)
 
-        if prev_state["runner_config"]["runner_storage"] != runner_storage:
-            logger.error(
-                "Storage option changed from %s to %s, blocking the charm",
-                prev_state["runner_config"]["runner_storage"],
-                runner_storage,
-            )
-            raise ImmutableConfigChangedError(
-                msg="runner-storage config cannot be changed after deployment, redeploy if needed"
-            )
-        if prev_state["runner_config"]["base_image"] != base_image.value:
-            logger.error(
-                "Base image option changed from %s to %s, blocking the charm",
-                prev_state["runner_config"]["base_image"],
-                runner_storage,
+        try:
+            if prev_state["runner_config"]["runner_storage"] != runner_storage:
+                logger.error(
+                    "Storage option changed from %s to %s, blocking the charm",
+                    prev_state["runner_config"]["runner_storage"],
+                    runner_storage,
+                )
+                raise ImmutableConfigChangedError(
+                    msg=(
+                        "runner-storage config cannot be changed after deployment, "
+                        "redeploy if needed"
+                    )
+                )
+        except KeyError as exc:
+            logger.warning(
```

Comment:
What is the reason that this is a warning please? Isn't it entirely expected if we upgrade from a previous version that didn't have the key?

In general, the way I think about warnings is that we should have some intention of making this an error later. So use cases include that we warn that a function is deprecated and will be removed in the future. Or that a retry has been triggered twice and, if it fails again, that it will result in an error. For whoever is reading the warning in the log there should be some action, otherwise what are we warning about? Warnings shouldn't be routinely ignored

Reply:
I think this warning doesn't happen routinely - only during charm upgrades with backward-incompatibilities. Hence I think it is worth noting that this action is taking place.
Would you think this would then be more suitable for a debug/info log?

Reply:
yeh info seems appropriate.

What is meant by routinely is that every time it would have appeared as a warning there is no action to be taken by the operator, hence they would ignore it

Reply:
Updated the code